File: sdk/steerplane/policy_engine.py
"""
SteerPlane SDK — Policy Engine

Rule-based action control for agent runs.
Supports allow/deny lists (with glob patterns), per-action rate limiting,
and human-in-the-loop approval workflows.
"""

import time
import fnmatch
import logging
from dataclasses import dataclass, field
from typing import Callable, Optional

from .exceptions import PolicyViolationError

logger = logging.getLogger("steerplane")


@dataclass
class PolicyDecision:
    """Result of a policy check."""
    allowed: bool
    action: str
    rule: str = ""       # Which rule matched (e.g. "deny:delete_*")
    reason: str = ""     # Human-readable explanation


@dataclass
class RateLimitSpec:
    """Rate limit specification for an action pattern."""
    pattern: str
    max_count: int
    window_seconds: float


class PolicyEngine:
    """
    Evaluates incoming actions against a set of policy rules.

    Rules are checked in order:
      1. Deny list (block immediately)
      2. Allow list (if non-empty, action must match at least one pattern)
      3. Rate limits (sliding-window counters per action pattern)
      4. Approval workflow (calls callback; block if denied or no callback)

    Args:
        allowed_actions: Glob patterns for permitted actions. If empty, all
            actions are permitted (unless denied).
        denied_actions: Glob patterns for forbidden actions.
        rate_limits: List of RateLimitSpec or dicts with keys
            ``pattern``, ``max_count``, ``window_seconds``.
        require_approval: Glob patterns requiring human approval.
        approval_callback: ``(action, metadata) → bool`` called when an
            action matches a require_approval pattern.
    """

    def __init__(
        self,
        allowed_actions: list[str] | None = None,
        denied_actions: list[str] | None = None,
        rate_limits: list[RateLimitSpec | dict] | None = None,
        require_approval: list[str] | None = None,
        approval_callback: Callable[[str, dict | None], bool] | None = None,
    ):
        self.allowed_actions: list[str] = allowed_actions or []
        self.denied_actions: list[str] = denied_actions or []
        self.require_approval: list[str] = require_approval or []
        self.approval_callback = approval_callback

        # Normalise rate-limit specs
        self.rate_limits: list[RateLimitSpec] = []
        for rl in (rate_limits or []):
            if isinstance(rl, dict):
                self.rate_limits.append(RateLimitSpec(**rl))
            else:
                self.rate_limits.append(rl)

        # Sliding-window timestamps keyed by pattern
        self._rate_windows: dict[str, list[float]] = {
            rl.pattern: [] for rl in self.rate_limits
        }

    # ────────────────────── public API ──────────────────────

    def check(self, action: str, metadata: dict | None = None) -> PolicyDecision:
        """
        Evaluate *action* against all policy rules.

        Returns a ``PolicyDecision`` on success.
        Raises ``PolicyViolationError`` if the action is denied.
        """
        # 1. Deny list — checked first, always wins
        for pattern in self.denied_actions:
            if fnmatch.fnmatch(action, pattern):
                self._deny(action, rule=f"deny:{pattern}",
                           reason=f"Action '{action}' matches deny pattern '{pattern}'")

        # 2. Allow list — if non-empty, action MUST match at least one entry
        if self.allowed_actions:
            if not any(fnmatch.fnmatch(action, p) for p in self.allowed_actions):
                self._deny(action, rule="allow_list",
                           reason=f"Action '{action}' is not in the allow list")

        # 3. Rate limits
        now = time.monotonic()
        for rl in self.rate_limits:
            if fnmatch.fnmatch(action, rl.pattern):
                window = self._rate_windows[rl.pattern]
                # Prune expired timestamps
                cutoff = now - rl.window_seconds
                window[:] = [t for t in window if t > cutoff]

                if len(window) >= rl.max_count:
                    self._deny(
                        action,
                        rule=f"rate_limit:{rl.pattern}",
                        reason=(
                            f"Rate limit exceeded for '{action}': "
                            f"{rl.max_count} calls per {rl.window_seconds}s"
                        ),
                    )
                # Record this call
                window.append(now)

        # 4. Approval workflow
        for pattern in self.require_approval:
            if fnmatch.fnmatch(action, pattern):
                if self.approval_callback is None:
                    self._deny(
                        action,
                        rule=f"approval:{pattern}",
                        reason=f"Action '{action}' requires approval but no callback is set",
                    )
                elif not self.approval_callback(action, metadata):
                    self._deny(
                        action,
                        rule=f"approval:{pattern}",
                        reason=f"Action '{action}' was denied by the approval callback",
                    )

        return PolicyDecision(allowed=True, action=action)

    @property
    def has_rules(self) -> bool:
        """True if at least one rule is configured."""
        return bool(
            self.allowed_actions
            or self.denied_actions
            or self.rate_limits
            or self.require_approval
        )

    # ────────────────────── internals ──────────────────────

    @staticmethod
    def _deny(action: str, rule: str, reason: str):
        """Raise a PolicyViolationError."""
        raise PolicyViolationError(action=action, rule=rule, reason=reason)

File: sdk-ts/src/policy-engine.ts
/**
 * SteerPlane SDK — Policy Engine (TypeScript)
 *
 * Rule-based action control for agent runs.
 * Supports allow/deny lists (with glob patterns), per-action rate limiting,
 * and human-in-the-loop approval workflows.
 */

import { PolicyViolationError } from "./errors.js";

/** Result of a policy check. */
export interface PolicyDecision {
  allowed: boolean;
  action: string;
  /** Which rule matched (e.g. "deny:delete_*"). */
  rule: string;
  /** Human-readable explanation. */
  reason: string;
}

/** Rate limit specification for an action pattern. */
export interface RateLimitSpec {
  /** Glob pattern to match against action names. */
  pattern: string;
  /** Maximum number of calls allowed within the window. */
  maxCount: number;
  /** Sliding window duration in seconds. */
  windowSeconds: number;
}

/** Approval callback type. Return true to allow, false to deny. */
export type ApprovalCallback = (
  action: string,
  metadata?: Record<string, unknown>
) => boolean | Promise<boolean>;

/** Options to construct a PolicyEngine. */
export interface PolicyEngineOptions {
  /** Glob patterns for permitted actions. If empty, all are permitted (unless denied). */
  allowedActions?: string[];
  /** Glob patterns for forbidden actions. */
  deniedActions?: string[];
  /** Per-action rate limits. */
  rateLimits?: RateLimitSpec[];
  /** Glob patterns requiring human approval before execution. */
  requireApproval?: string[];
  /** Callback invoked when an action requires approval. */
  approvalCallback?: ApprovalCallback;
}

/**
 * Evaluates incoming actions against a set of policy rules.
 *
 * Rules are checked in order:
 *   1. Deny list (block immediately)
 *   2. Allow list (if non-empty, action must match at least one pattern)
 *   3. Rate limits (sliding-window counters per action pattern)
 *   4. Approval workflow (calls callback; block if denied or no callback)
 */
export class PolicyEngine {
  public readonly allowedActions: string[];
  public readonly deniedActions: string[];
  public readonly rateLimits: RateLimitSpec[];
  public readonly requireApproval: string[];
  public approvalCallback?: ApprovalCallback;

  private readonly _rateWindows: Map<string, number[]> = new Map();

  constructor(opts: PolicyEngineOptions = {}) {
    this.allowedActions = opts.allowedActions ?? [];
    this.deniedActions = opts.deniedActions ?? [];
    this.rateLimits = opts.rateLimits ?? [];
    this.requireApproval = opts.requireApproval ?? [];
    this.approvalCallback = opts.approvalCallback;

    // Initialise sliding-window arrays
    for (const rl of this.rateLimits) {
      this._rateWindows.set(rl.pattern, []);
    }
  }

  /**
   * Evaluate *action* against all policy rules.
   *
   * Returns a `PolicyDecision` on success.
   * Throws `PolicyViolationError` if the action is denied.
   */
  async check(
    action: string,
    metadata?: Record<string, unknown>
  ): Promise<PolicyDecision> {
    // 1. Deny list — checked first, always wins
    for (const pattern of this.deniedActions) {
      if (globMatch(action, pattern)) {
        PolicyEngine._deny(
          action,
          `deny:${pattern}`,
          `Action '${action}' matches deny pattern '${pattern}'`
        );
      }
    }

    // 2. Allow list — if non-empty, action MUST match at least one entry
    if (this.allowedActions.length > 0) {
      if (!this.allowedActions.some((p) => globMatch(action, p))) {
        PolicyEngine._deny(
          action,
          "allow_list",
          `Action '${action}' is not in the allow list`
        );
      }
    }

    // 3. Rate limits
    const now = performance.now() / 1000; // seconds (monotonic-like)
    for (const rl of this.rateLimits) {
      if (globMatch(action, rl.pattern)) {
        const window = this._rateWindows.get(rl.pattern)!;
        // Prune expired timestamps
        const cutoff = now - rl.windowSeconds;
        const pruned = window.filter((t) => t > cutoff);
        this._rateWindows.set(rl.pattern, pruned);

        if (pruned.length >= rl.maxCount) {
          PolicyEngine._deny(
            action,
            `rate_limit:${rl.pattern}`,
            `Rate limit exceeded for '${action}': ` +
              `${rl.maxCount} calls per ${rl.windowSeconds}s`
          );
        }
        // Record this call
        pruned.push(now);
      }
    }

    // 4. Approval workflow
    for (const pattern of this.requireApproval) {
      if (globMatch(action, pattern)) {
        if (!this.approvalCallback) {
          PolicyEngine._deny(
            action,
            `approval:${pattern}`,
            `Action '${action}' requires approval but no callback is set`
          );
        } else {
          const approved = await this.approvalCallback(action, metadata);
          if (!approved) {
            PolicyEngine._deny(
              action,
              `approval:${pattern}`,
              `Action '${action}' was denied by the approval callback`
            );
          }
        }
      }
    }

    return { allowed: true, action, rule: "", reason: "" };
  }

  /** True if at least one rule is configured. */
  get hasRules(): boolean {
    return (
      this.allowedActions.length > 0 ||
      this.deniedActions.length > 0 ||
      this.rateLimits.length > 0 ||
      this.requireApproval.length > 0
    );
  }

  /** Throw PolicyViolationError. */
  private static _deny(action: string, rule: string, reason: string): never {
    throw new PolicyViolationError(action, rule, reason);
  }
}

// ────────────────── Glob matching (minimal fnmatch) ──────────────────

/**
 * Simple glob match implementation supporting `*` and `?` wildcards.
 * Mirrors Python's `fnmatch.fnmatch` behaviour.
 */
export function globMatch(text: string, pattern: string): boolean {
  // Convert glob pattern to regex
  let regex = "^";
  for (const ch of pattern) {
    switch (ch) {
      case "*":
        regex += ".*";
        break;
      case "?":
        regex += ".";
        break;
      case ".":
      case "+":
      case "^":
      case "$":
      case "(":
      case ")":
      case "{":
      case "}":
      case "[":
      case "]":
      case "|":
      case "\\":
        regex += `\\${ch}`;
        break;
      default:
        regex += ch;
    }
  }
  regex += "$";
  return new RegExp(regex, "i").test(text);
}
